x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -0,0 +1,158 @@
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+/**
+ * Inspects all aliases that have greater privileges than the indices that they point to and logs the role descriptor, granting privileges
+ * in this manner, as deprecated and requiring changes. This is done in preparation for the removal of the ability to define privileges over
+ * aliases. The log messages are generated asynchronously and do not generate deprecation response headers. One log entry is generated for
+ * each role descriptor and alias pair, and it contains all the indices for which privileges are a subset of those of the alias. In this
+ * case, the administrator has to adjust the index privileges definition of the respective role such that name patterns do not cover aliases
+ * (or rename aliases). If no logging is generated then the roles used for the current indices and aliases are not vulnerable to the
+ * subsequent breaking change. However, there could be role descriptors that are not used (not mapped to a user that is currently using the
+ * system) which are invisible to this check. Moreover, role descriptors can be dynamically added by role providers. In addition, role
+ * descriptors are merged when building the effective role, so a role-alias pair reported as deprecated might not actually have an impact if
+ * other role descriptors cover its indices. The check iterates over all indices and aliases for each role descriptor so it is quite
+ * expensive computationally. For this reason the check is done only once a day for each role. If the role definitions stay the same, the
+ * deprecations can change from one day to another only if aliases or indices are added.
+ */
+public final class DeprecationRoleDescriptorConsumer implements Consumer<Collection<RoleDescriptor>> {
+
+    private static final String ROLE_PERMISSION_DEPRECATION_STANZA = "Role [%s] contains index privileges covering the [%s] alias but"
+            + " which do not cover some of the indices that it points to [%s]. Granting privileges over an alias and hence granting"
+            + " privileges over all the indices that the alias points to is deprecated and will be removed in a future version of"
+            + " Elasticsearch. Instead define permissions exclusively on index names or index name patterns.";
+
+    private final DeprecationLogger deprecationLogger;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final Set<String> dailyRoleCache;
+
+    public DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
```

Comment:
I would rather have a public constructor without the deprecation logger that calls a package private one that takes the deprecation logger. This way we can use it for testing and we don't take the logger from the `Security` class.